### PR TITLE
working suggestion box for ayahs but i want it disabled on the surah …

### PR DESCRIPTION
**Pull request by sharabash**
…page; the css is inline and the look is not so good, i still need to make a css file; i also need to feed in surah name suggestions somehow;

@mmahalwy i'd like your feedback, if you can check this out and tell me what needs to happen

### CODE REVIEW

**Comment by sharabash** [comment]
@mmahalwy both branches (here and in -api) are in a "done" stage and ready for review now. the only outstanding issue I see for the time being is that I want the autocomplete component to *not* be included if we're on the surah page. how do I do that?

**Comment by mmahalwy** [comment]
I'd wait till #221 is done.

File: src/scripts/components/header/SearchAutocomplete.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import classNames from 'classnames';
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: don't nee

File: src/scripts/components/header/SearchAutocomplete.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import SurahsStore from 'stores/SurahsStore';
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: :scissors:

File: src/scripts/components/header/SearchAutocomplete.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import connectToStores from 'fluxible-addons-react/connectToStores';
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: :scissors:

File: src/scripts/components/header/SearchAutocomplete.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import request from 'superagent';
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: when you rebase, you can now use `ApiClient`

File: src/scripts/components/header/SearchAutocomplete.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import Settings from 'constants/Settings';
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: cut

File: src/scripts/components/header/SearchAutocomplete.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+class SearchAutocomplete extends React.Component {
+  constructor(props, context) {
+    super(props, context);
+    this.state = { value: '', ayat: [], surahs: [] };
+    this.cached = {};
+  }
+  componentWillReceiveProps( nextProps ) {
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: i don't think this is needed.

File: src/scripts/components/header/SearchAutocomplete.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      this.handleSurahSuggestions(value);
+      if (value.length >= 3) {
+        this.handleAyahSuggestions(value);
+      }
+      else {
+        this.setState({ ayat: [] });
+      }
+    }
+    else {
+      this.setState( { surahs: [] } );
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: not a fan of setting state every key down, even if not needed. It should set state if length less than 3 & `this.state.surahs.length`

File: src/scripts/components/header/SearchAutocomplete.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  };
+
+  handleSurahSuggestions(value) {
+    var escaped = value.replace(/[-\\^$*+?.()|[\]{}]/g, "\\$&");
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: `const` not `var`

File: src/scripts/components/header/SearchAutocomplete.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    var matches = [];
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: `const`

File: src/scripts/components/header/SearchAutocomplete.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    this.props.surahs.forEach( function( surah ) {
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: use arrow function: `(surah) => {}` and no spaces before

File: src/scripts/components/header/SearchAutocomplete.js [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+      if ( RegExp( escaped, "i" ).test( surah.name.simple.replace( /['-]/g, '' )  ) ) {
+        matches.push( [ surah.name.simple, surah.id ] );
+      } else
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: this is weird

File: src/scripts/components/header/SearchAutocomplete.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+      if ( RegExp( escaped, "i" ).test( surah.name.arabic ) ) {
+        matches.push( [ surah.name.arabic, surah.id ] );
+      }
+    } );
+    matches = matches.sort(function( a, b ) {
+      return a[1] < b[1] ? -1 : 1; // <-- order by surah id
+      return RegExp( '^'+ escaped, 'i' ).test( a[0] ) ? -1 : 1; // <-- order by whether it starts the same
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: it'll never make it to this line...

File: src/scripts/components/header/SearchAutocomplete.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    }).map( function( s ) {
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: whats `s`?

File: src/scripts/components/header/SearchAutocomplete.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+      // var text = s[0].replace(RegExp(escaped, "gi"), "<mark>$&</mark>"); // <-- this will highlight the matched part
+      var text = s[0];
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: `const`

File: src/scripts/components/header/SearchAutocomplete.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+      } else
+      if ( RegExp( escaped, "i" ).test( surah.name.arabic ) ) {
+        matches.push( [ surah.name.arabic, surah.id ] );
+      }
+    } );
+    matches = matches.sort(function( a, b ) {
+      return a[1] < b[1] ? -1 : 1; // <-- order by surah id
+      return RegExp( '^'+ escaped, 'i' ).test( a[0] ) ? -1 : 1; // <-- order by whether it starts the same
+    }).map( function( s ) {
+      // var text = s[0].replace(RegExp(escaped, "gi"), "<mark>$&</mark>"); // <-- this will highlight the matched part
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: :scissors:

File: src/scripts/components/header/SearchAutocomplete.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      var text = s[0];
+      return { text: '<b>'+ text +'</b>', href: '/'+ s[1] };
+    } ).slice(0,5);
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: `.slice(0, 5)

File: src/scripts/components/header/SearchAutocomplete.js [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    this.setState( { surahs: matches } );
+  };
+
+  handleAyahSuggestions(value) {
+    var self = this;
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: you won't need `self` if you use arrow functions at the bottom. So `.end((err, res) => {})`

File: src/scripts/components/header/SearchAutocomplete.js [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+  renderList(key) {
+    return this.state[key].map((item) => {
+      return this.renderItem(item);
+    });
+  };
+
+  renderItem(item) {
+    return (
+      <li key={item.href}>
+        <div>
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: why?

File: src/scripts/components/header/SearchAutocomplete.js [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+      <div className={className}>
+        <ul role="menu">
+          {this.renderList('surahs')}
+          {this.renderList('ayat')}
+        </ul>
+      </div>
+    );
+  }
+}
+SearchAutocomplete.contextTypes = {
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: :scissors:

File: src/scripts/components/header/SearchAutocomplete.js [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+  executeAction: React.PropTypes.func.isRequired
+};
+SearchAutocomplete = connectToStores( SearchAutocomplete, [SurahsStore], (context, props) => {
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: :scissors:

File: src/styles/components/SearchAutocomplete.scss [hunk 21]
```diff
@@ -0,0 +1,93 @@
+.search-autocomplete {
```
Comment (mmahalwy) [inline src/styles/components/SearchAutocomplete.scss]: Id recommend using the css-modules way of css. See: https://github.com/quran/quran.com-frontend/blob/master/src/containers/Home/index.js#L11

and file: https://github.com/quran/quran.com-frontend/blob/master/src/containers/Home/style.scss

File: src/scripts/components/header/SearchAutocomplete.js [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+  };
+
+  render() {
+    var className = classNames({
+      'search-autocomplete': true,
+      'hidden': this.state.ayat.length || this.state.surahs.length ? false : true
+    });
+
+    return (
+      <div className={className}>
```
Comment (mmahalwy) [inline src/scripts/components/header/SearchAutocomplete.js]: ``search-autocomplete ${this.state.ayat.length || this.state.surahs.length ? '' : 'hidden'}``

**Comment by mmahalwy** [comment]
For some reason eslint is not running, otherwise it'd be very mad at you. We need to keep the spacing consistent. So, for example, `someFn( function() {} );` is not allowed. should be `someFn(function() {});`. This is also the case for `if` statements.

File: src/scripts/components/header/SearchAutocomplete.js (hunk 1 above)
Comment (sharabash) [inline src/scripts/components/header/SearchAutocomplete.js]: how do i not need React?

File: src/scripts/components/header/SearchAutocomplete.js (hunk 18 above)
Comment (sharabash) [inline src/scripts/components/header/SearchAutocomplete.js]: why what?

File: src/scripts/components/header/SearchAutocomplete.js (hunk 15 above)
Comment (sharabash) [inline src/scripts/components/header/SearchAutocomplete.js]: cut what?

File: src/scripts/components/header/SearchAutocomplete.js (hunk 14 above)
Comment (sharabash) [inline src/scripts/components/header/SearchAutocomplete.js]: whats a const

File: src/scripts/components/header/SearchAutocomplete.js (hunk 11 above)
Comment (sharabash) [inline src/scripts/components/header/SearchAutocomplete.js]: what is weird

**Pull request closed without merging**